In [13]:
# pip install mlxtend

  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 0.23.1
    Uninstalling scikit-learn-0.23.1:
      Successfully uninstalled scikit-learn-0.23.1
Note: you may need to restart the kernel to use updated packages.


ERROR: scikit-learn 1.2.0 has requirement joblib>=1.1.1, but you'll have joblib 0.16.0 which is incompatible.


In [40]:
from csv import reader
from mlxtend.preprocessing import TransactionEncoder
import pandas as pd
from mlxtend.frequent_patterns import apriori, fpgrowth, fpmax, association_rules
import numpy as np


In [10]:
groceries = []

with open('data/groceries.csv', 'r', encoding='utf-8-sig') as csvfile:
    csv_reader = reader(csvfile)
    for row in csv_reader:
        groceries.append(row)

In [12]:
groceries[0:5]

[['citrus fruit', 'semi-finished bread', 'margarine', 'ready soups'],
 ['tropical fruit', 'yogurt', 'coffee'],
 ['whole milk'],
 ['pip fruit', 'yogurt', 'cream cheese', 'meat spreads'],
 ['other vegetables',
  'whole milk',
  'condensed milk',
  'long life bakery product']]

In [17]:
# Instantiate an encoder object
encoder = TransactionEncoder()
# Use the fit method to extract unique labels in the set
# Use the transform method to one-hot encode the transactions into a boolean
transactions = encoder.fit(groceries).transform(groceries)
transactions

array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False,  True, False],
       [False, False, False, ...,  True, False, False],
       ...,
       [False, False, False, ..., False,  True, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False]])

In [19]:
# Convert the transactions to a dataframe
itemsets = pd.DataFrame(transactions, columns = encoder.columns_)
itemsets.head()

abrasive cleaner  artif. sweetener  ...  yogurt  zwieback
0             False             False  ...   False     False
1             False             False  ...    True     False
2             False             False  ...   False     False
3             False             False  ...    True     False
4             False             False  ...   False     False

[5 rows x 169 columns]

In [20]:
# Summary of the dataframe
itemsets.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9835 entries, 0 to 9834
Columns: 169 entries, abrasive cleaner to zwieback
dtypes: bool(169)
memory usage: 1.6 MB


# Generate frequent itemsets

Assume we want itemsets that occur at least 5 times a day. The data spans 30 days and has 9,835 transactions. We need to set our minimum support threshold to $5 \times \frac{30}{9835} \approx 0.015$

In [23]:
# Display the frequent itemsets with column names
frequent_itemsets = apriori(itemsets, min_support = 0.015, use_colnames = True)
frequent_itemsets

support                                         itemsets
0    0.017692                                  (baking powder)
1    0.052466                                           (beef)
2    0.033249                                        (berries)
3    0.026029                                      (beverages)
4    0.080529                                   (bottled beer)
..        ...                                              ...
175  0.023183  (other vegetables, root vegetables, whole milk)
176  0.017082   (other vegetables, whole milk, tropical fruit)
177  0.022267           (other vegetables, yogurt, whole milk)
178  0.015557                 (yogurt, whole milk, rolls/buns)
179  0.015150             (yogurt, whole milk, tropical fruit)

[180 rows x 2 columns]

In [25]:
# Sort by support
frequent_itemsets.sort_values('support', ascending = False)

support                              itemsets
71   0.255516                          (whole milk)
44   0.193493                    (other vegetables)
53   0.183935                          (rolls/buns)
60   0.174377                                (soda)
72   0.139502                              (yogurt)
..        ...                                   ...
163  0.015252               (shopping bags, yogurt)
179  0.015150  (yogurt, whole milk, tropical fruit)
45   0.015048                               (pasta)
11   0.015048                         (canned fish)
167  0.015048                   (whole milk, sugar)

[180 rows x 2 columns]

In [26]:
# Let's look at only frequent itemsets with a length greater than two. First get the length of the itemsets:
length = frequent_itemsets['itemsets'].str.len()
# Then create a filter:
rows = length > 2
# Finally, apply the filter to the dataframe:
frequent_itemsets[rows]

support                                         itemsets
174  0.017895       (other vegetables, whole milk, rolls/buns)
175  0.023183  (other vegetables, root vegetables, whole milk)
176  0.017082   (other vegetables, whole milk, tropical fruit)
177  0.022267           (other vegetables, yogurt, whole milk)
178  0.015557                 (yogurt, whole milk, rolls/buns)
179  0.015150             (yogurt, whole milk, tropical fruit)

In [27]:
frequent_itemsets.groupby(length)['support'].describe()

count      mean       std  ...       50%       75%       max
itemsets                             ...                              
1          73.0  0.053441  0.045956  ...  0.037112  0.064870  0.255516
2         101.0  0.024799  0.010058  ...  0.021047  0.027555  0.074835
3           6.0  0.018522  0.003417  ...  0.017489  0.021174  0.023183

[3 rows x 8 columns]

Most transactions considered frequent are two-item purchases (count: 101). Single-item purchases have a higher support value (according to the mean and 50% columns) than two- and three-item purchases.

# Create association rules

In [30]:
# Generate rules - looking for confidence with a minimum threshold of 0.25
rules = association_rules(frequent_itemsets, metric = 'confidence', min_threshold = 0.25)
rules

antecedents         consequents  ...  leverage  conviction
0                         (beef)  (other vegetables)  ...  0.009574    1.292416
1                         (beef)   (root vegetables)  ...  0.011668    1.332628
2                         (beef)        (whole milk)  ...  0.007845    1.251315
3                 (bottled beer)        (whole milk)  ... -0.000139    0.997684
4                (bottled water)              (soda)  ...  0.009705    1.119017
..                           ...                 ...  ...       ...         ...
73          (yogurt, rolls/buns)        (whole milk)  ...  0.006775    1.360192
74      (whole milk, rolls/buns)            (yogurt)  ...  0.007656    1.186380
75          (yogurt, whole milk)    (tropical fruit)  ...  0.009271    1.226823
76      (yogurt, tropical fruit)        (whole milk)  ...  0.007668    1.542528
77  (whole milk, tropical fruit)            (yogurt)  ...  0.009249    1.340701

[78 rows x 9 columns]

In [31]:
# Filter to see only 'rolls/buns' in the antecedent
rows = rules['antecedents'] == {'rolls/buns'}
# Apply the filter
rules[rows]

antecedents   consequents  ...  leverage  conviction
51  (rolls/buns)  (whole milk)  ...  0.009636    1.075696

[1 rows x 9 columns]

In [32]:
# Repeat the process for the consequent
rows = rules['consequents'] == {'rolls/buns'}
rules[rows]

antecedents   consequents  ...  leverage  conviction
24         (frankfurter)  (rolls/buns)  ...  0.008370    1.210531
50             (sausage)  (rolls/buns)  ...  0.013324    1.210344
72  (yogurt, whole milk)  (rolls/buns)  ...  0.005252    1.129779

[3 rows x 9 columns]

In [33]:
# Filter to see all rules with 'rolls/buns' as either the antecedent or consequent
rows = rules['antecedents'].astype(str).str.contains('rolls/buns')
rules[rows]

antecedents         consequents  ...  leverage  conviction
51                    (rolls/buns)        (whole milk)  ...  0.009636    1.075696
62  (other vegetables, rolls/buns)        (whole milk)  ...  0.007010    1.283699
63        (whole milk, rolls/buns)  (other vegetables)  ...  0.006937    1.179067
73            (yogurt, rolls/buns)        (whole milk)  ...  0.006775    1.360192
74        (whole milk, rolls/buns)            (yogurt)  ...  0.007656    1.186380

[5 rows x 9 columns]

In [34]:
# Filter for rules with a length of 2 or more
rows = rules['antecedents'].str.len() > 1
rules[rows]

antecedents  ... conviction
62       (other vegetables, rolls/buns)  ...   1.283699
63             (whole milk, rolls/buns)  ...   1.179067
64  (other vegetables, root vegetables)  ...   1.457687
65       (other vegetables, whole milk)  ...   1.290900
66        (root vegetables, whole milk)  ...   1.533320
67   (other vegetables, tropical fruit)  ...   1.420556
68         (whole milk, tropical fruit)  ...   1.352851
69           (other vegetables, yogurt)  ...   1.528340
70       (other vegetables, whole milk)  ...   1.225003
71                 (yogurt, whole milk)  ...   1.338511
72                 (yogurt, whole milk)  ...   1.129779
73                 (yogurt, rolls/buns)  ...   1.360192
74             (whole milk, rolls/buns)  ...   1.186380
75                 (yogurt, whole milk)  ...   1.226823
76             (yogurt, tropical fruit)  ...   1.542528
77         (whole milk, tropical fruit)  ...   1.340701

[16 rows x 9 columns]

In [35]:
# Let's look for a lift of more than 2, leverage score more than 0.01, and a conviction score of more than 1.4
rows = ((rules['lift'] > 2) & (rules['leverage'] > 0.01) & (rules['conviction'] > 1.4))
rules[rows]

antecedents         consequents  ...  leverage  conviction
39              (root vegetables)  (other vegetables)  ...  0.026291    1.426693
66  (root vegetables, whole milk)  (other vegetables)  ...  0.013719    1.533320
69     (other vegetables, yogurt)        (whole milk)  ...  0.011174    1.528340

[3 rows x 9 columns]

# Evaluate association rules

In [36]:
rules.describe()

antecedent support  consequent support  ...   leverage  conviction
count           78.000000           78.000000  ...  78.000000   78.000000
mean             0.073186            0.211041  ...   0.010156    1.245113
std              0.036738            0.046866  ...   0.004766    0.115688
min              0.029283            0.104931  ...  -0.000139    0.997684
25%              0.052466            0.193493  ...   0.006970    1.166832
50%              0.061210            0.193493  ...   0.009260    1.226608
75%              0.082766            0.255516  ...   0.011788    1.294081
max              0.255516            0.255516  ...   0.026291    1.542528

[8 rows x 7 columns]

In [37]:
# Explore the top 5 rules by lift
rules.sort_values('lift', ascending = False).head()

antecedents        consequents  ...  leverage  conviction
1                           (beef)  (root vegetables)  ...  0.011668    1.332628
65  (other vegetables, whole milk)  (root vegetables)  ...  0.015026    1.290900
75            (yogurt, whole milk)   (tropical fruit)  ...  0.009271    1.226823
47                     (pip fruit)   (tropical fruit)  ...  0.012499    1.226392
77    (whole milk, tropical fruit)           (yogurt)  ...  0.009249    1.340701

[5 rows x 9 columns]

In [38]:
# The lift values above are above 1, meaning we consider it a factor of likelihood (the first rule shows that customers who bought beef are 3.04 times more likely to purchase root vegetables).

# Explore the top 5 rules by leverage
rules.sort_values('leverage', ascending = False).head()

antecedents         consequents  ...  leverage  conviction
39   (root vegetables)  (other vegetables)  ...  0.026291    1.426693
43  (other vegetables)        (whole milk)  ...  0.025394    1.214013
44        (whole milk)  (other vegetables)  ...  0.025394    1.140548
52   (root vegetables)        (whole milk)  ...  0.021056    1.350401
61            (yogurt)        (whole milk)  ...  0.020379    1.244132

[5 rows x 9 columns]

In [39]:
# These rules seem to be in conflict with each other (or at least deceiving). Maybe the consequent and antecedent are not dependent on each other?

# Let's explore the top 5 rules by conviction
rules.sort_values('conviction', ascending = False).head()

antecedents         consequents  ...  leverage  conviction
76       (yogurt, tropical fruit)        (whole milk)  ...  0.007668    1.542528
66  (root vegetables, whole milk)  (other vegetables)  ...  0.013719    1.533320
69     (other vegetables, yogurt)        (whole milk)  ...  0.011174    1.528340
9                        (butter)        (whole milk)  ...  0.013395    1.480817
19                         (curd)        (whole milk)  ...  0.012517    1.461085

[5 rows x 9 columns]

In [43]:
# Calculate Zhang's metric and add it to the rules dataframe

def zhang_metric(rules):
    sup = rules['support'].copy()
    sup_a = rules['antecedent support'].copy()
    sup_b = rules['consequent support'].copy()
    num = sup - sup_a * sup_b
    denom = np.max((sup * (1 - sup_a).values, sup_a * (sup_b - sup).values), axis = 0)
    return num / denom

In [44]:
# Assign the function output to a new column
rules['zhang'] = zhang_metric(rules)
rules

antecedents         consequents  ...  conviction     zhang
0                         (beef)  (other vegetables)  ...    1.292416  0.512224
1                         (beef)   (root vegetables)  ...    1.332628  0.708251
2                         (beef)        (whole milk)  ...    1.251315  0.389597
3                 (bottled beer)        (whole milk)  ...    0.997684 -0.007351
4                (bottled water)              (soda)  ...    1.119017  0.376535
..                           ...                 ...  ...         ...       ...
73          (yogurt, rolls/buns)        (whole milk)  ...    1.360192  0.451027
74      (whole milk, rolls/buns)            (yogurt)  ...    1.186380  0.521686
75          (yogurt, whole milk)    (tropical fruit)  ...    1.226823  0.648285
76      (yogurt, tropical fruit)        (whole milk)  ...    1.542528  0.521384
77  (whole milk, tropical fruit)            (yogurt)  ...    1.340701  0.637483

[78 rows x 10 columns]

In [45]:
# Consider the top 5 by Zhang metric
rules.sort_values('zhang', ascending = False).head()

antecedents        consequents  ...  conviction     zhang
1                           (beef)  (root vegetables)  ...    1.332628  0.708251
65  (other vegetables, whole milk)  (root vegetables)  ...    1.290900  0.700572
47                     (pip fruit)   (tropical fruit)  ...    1.226392  0.661650
75            (yogurt, whole milk)   (tropical fruit)  ...    1.226823  0.648285
77    (whole milk, tropical fruit)           (yogurt)  ...    1.340701  0.637483

[5 rows x 10 columns]

In [46]:
# The above have a strong positive association. 

# Consider the bottom 5 by Zhang metric
rules.sort_values('zhang', ascending = True).head()

antecedents   consequents  ...  conviction     zhang
3    (bottled beer)  (whole milk)  ...    0.997684 -0.007351
5   (bottled water)  (whole milk)  ...    1.080446  0.200417
51     (rolls/buns)  (whole milk)  ...    1.075696  0.208496
54        (sausage)  (whole milk)  ...    1.091910  0.217372
16         (coffee)  (whole milk)  ...    1.098451  0.219830

[5 rows x 10 columns]

In [ ]:
# This first rule has a negative score, indicating a slight dissociation. Meaning, moving bottled beer and whole milk away from each other would not have a significant impact, nor would it be a good idea to pair them together for a promotional sale.